### PR TITLE
Enable Must guarding in aggregate root

### PR DESCRIPTION
**Pull request by Corniel**
As seen in the [Conquer Club example](https://github.com/Qowaiv/qowaiv-domainmodel/tree/master/example), having guards/pre-condition checks on methods that change the state of the aggregate is a powerful way to build an anti-corruption layer, and/or guarantee the aggregate to stay valid.

With this pull request, this `Must` syntax becomes a first-class citizen of the aggregate. Out of the box, is has a `Be`, `NotBe`, and a `Exist` method. By making the `TSubject Subject` property public, there should be no limitations, on writing extension methods on `Must<TSubject>` to have custom guarding; that is also the only way to extend, the class has been sealed.

``` C#
public Result<Game> Deploy(CountryId country, Army army) =>
    Must.BeInPhase(GamePhase.Deploy)
    | (g => g.Must.BeActivePlayer(army.Owner))
    | (g => g.Must.Exist(country))
    | (g => g.Must.BeOwnedBy(country, army.Owner))
    | (g => g.Must.NotExceedArmyBuffer(army))
    | (g => g.ApplyEvent(new Deployed(country, army)));
```

The `.Must()` functionality is defined in the Qowaiv.Validation.Guarding package.

### CODE REVIEW

**Review by mars888 — APPROVED** [review]
Interesting techniques. I quite like it 😄
Not sure if `Must` is a monad per se though.

**Comment by Corniel** [comment]
> 
> 
> Interesting techniques. I quite like it 😄
> Not sure if `Must` is a monad per se though.

According to [Wikipedia](https://en.wikipedia.org/wiki/Monad_(functional_programming))
> a monad is an abstraction that allows structuring programs generically (..) which represents a specific form of computation, along with two procedures (..)

And I think, that is what is does, or more precisely, the methods on the `Must` return a `Result<TSubject>` monad.

That being said, I'm still struggling with the place where to put it. An other logical place could be:

``` C#
namepace Qowaiv.Validation.Guarding
{
     public static Result<TModel> Must<TModel>(this TModel value) where TModel : class => new Must<TModel>(vaue);
} 
```
This should result in a separate package. I could even prevent direct creation of `Must<TSubject>`. Furthermore, doing so, reducing dependencies, the `Must<TAggregate> Must => this.Must();` property than could be dropped to reduce the dependencies required by `Qowaiv.DomainModel`. Hard choices...

File: test/Qowaiv.DomainModel.UnitTests/Must_specs.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        public void Must_Be_condition_is_met()
+        {
+            var model = new TestModel();
+            new Must<TestModel>(model).Be(condition: true, ValidationMessage.Error("This is wrong"))
+                .Should().BeValid().WithoutMessages()
+                .Value.Should().BeSameAs(model);
+        }
+
+        [Test]
+        public void Must_NotBe_condition_is_not_met()
```
Comment (LauraXiulan) [inline test/Qowaiv.DomainModel.UnitTests/Must_specs.cs:31]: This double negation is very confusing

Comment (Corniel) [reply test/Qowaiv.DomainModel.UnitTests/Must_specs.cs:31]: Any suggestion? `NotBe` is - in this case - defacto just a term, although I see your point.

**Comment by LauraXiulan** [comment]
I would put it in Validation.Messages. I don't think it warrants a separate package and a dependency on Validation.Messages in DomainModel is not that unreasonable

File: test/Qowaiv.DomainModel.UnitTests/Must_specs.cs (hunk 1 above)
Comment (LauraXiulan) [reply test/Qowaiv.DomainModel.UnitTests/Must_specs.cs:31]: Maybe, Must_NotBe_model_not_equal_to_condition? I find it hard to think of an easy to read title for this.

**Review by LauraXiulan — APPROVED** [review]

**Comment by Corniel** [comment]
> 
> 
> I would put it in Validation.Messages. I don't think it warrants a separate package and a dependency on Validation.Messages in DomainModel is not that unreasonable

But it has hardly anything to do with custom messages, don't you think that would be confusing?

It is an option not have a dependency in `Qowaiv.DomainModel`. `Must()` can be made available using an extension method.

**Comment by LauraXiulan** [comment]
> > I would put it in Validation.Messages. I don't think it warrants a separate package and a dependency on Validation.Messages in DomainModel is not that unreasonable
> 
> But it has hardly anything to do with custom messages, don't you think that would be confusing?
> 
> It is an option not have a dependency in `Qowaiv.DomainModel`. `Must()` can be made available using an extension method.

Isnt it, the Must is something that triggers the messages right? If not in Messages I would definitely put it in the Validation package.

File: test/Qowaiv.DomainModel.UnitTests/Must_specs.cs (hunk 1 above)
Comment (Corniel) [reply test/Qowaiv.DomainModel.UnitTests/Must_specs.cs:31]: if you put in `Math.Squared(4) != 2` the same happens. It is not model dependent per se.

**Comment by Corniel** [comment]
> Why not in Validation.Abstractions?

Validations is and should not be dependent on Messages. That would not allow `Must.Exist` to return a `EntityNotFound` message.

To me, it also does not feel as the **core** of `Qowaiv.Validations`, at least not to me.

**Comment by Corniel** [comment]
This pull request now only contains a Must property on the aggregate, and an extension on Must in the form of `HaveVersion(version)`. Obviously, the adjustments in the Conquer Club are relevant to show the usage.

File: src/Qowaiv.DomainModel/Qowaiv.Validation.Guarding/Musts.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        /// </typeparam>
+        /// <param name="must">
+        /// The must to extend on.
+        /// </param>
+        /// <param name="expected">
+        /// The expected version.
+        /// </param>
+        /// <remarks>
+        /// A dynamic is used, as the TId generic can not be resolved by usage.
+        /// It would be inconvenient to have specify the generics on this
```
Comment (LauraXiulan) [inline src/Qowaiv.DomainModel/Qowaiv.Validation.Guarding/Musts.cs:24]: to have **to**

**Review by LauraXiulan — APPROVED** [review]

**Pull request merged**